PVTPackage/source/MultiphaseSystem/PhaseSplitModel/BlackOilFlash.cpp

```diff
@@ -44,89 +44,105 @@ bool BlackOilFlash::computeEquilibrium( BlackOilFlashMultiphaseSystemProperties
-    sysProps.setWaterLnFugacity( gasAndWaterLnFugacity ); // FIXME It's the same values for water and gas?
+    // compositions
+    const std::vector< double > zeroMoleComposition{ 0.0, 0.0, 0.0 };
+    sysProps.setOilMoleComposition( zeroMoleComposition );
+    sysProps.setGasMoleComposition( zeroMoleComposition );
+    
+    // density, viscosity
+    BlackOilDeadOilProperties tmp( 0.0, 0.0, 0.0 );
+    sysProps.setOilModelProperties( tmp );
+    sysProps.setGasModelProperties( tmp );
+
+    // fugacities
+    sysProps.setOilLnFugacity( { std::log( 1.0 ), std::log( 1.0 ), std::log( 1.0 ) } );
+    sysProps.setGasLnFugacity( { std::log( 1.0 ), std::log( 1.0 ), std::log( 1.0 ) } );
+    sysProps.setWaterLnFugacity( { std::log( 1.0 ), std::log( 1.0 ), std::log( 1.0 ) } );        
+
+    // water
+    auto const waterProperties = m_waterPhaseModel.computeProperties( pressure );
+    sysProps.setWaterModelProperties( waterProperties );
+    
+    return true;
   }
-  else if( V > 1 ) //Only gas or undersaturated gas
-  {
-    LOGERROR( "Undersaturated gas not supported yet" );
-  }
-  else // Only oil or undersaturated oil
-  {
-    // Phase Fractions
-    sysProps.setOilFraction( 1. - zw );
-    sysProps.setGasFraction( 0. );
-    sysProps.setWaterFraction( zw );
-
+  else
+  {    
     // OIL
-    std::vector< double > const oilMoleComposition{ zo, zg, 0. }; // FIXME always 0.
-    sysProps.setOilMoleComposition( oilMoleComposition );
-    auto const oilUnderSaturatedProperties = m_oilPhaseModel.computeUnderSaturatedProperties( pressure, oilMoleComposition, gasSurfaceMoleDensity, gasSurfaceMassDensity );
-    sysProps.setOilModelProperties( oilUnderSaturatedProperties );
-
-    // LN FUGACITIES
-    sysProps.setOilLnFugacity( { std::log( 1. ), std::log( 1. ), std::log( 1. ) } );
-    sysProps.setGasLnFugacity( { std::log( 1. ), std::log( 1. ), std::log( 1. ) } );
-    sysProps.setWaterLnFugacity( { std::log( 1. ), std::log( 1. ), std::log( 1. ) } );
+    const double & oilSurfaceMoleDensity = m_oilPhaseModel.getSurfaceMoleDensity();
+    const double & oilSurfaceMassDensity = m_oilPhaseModel.getSurfaceMassDensity();
+    const double rsSat = m_oilPhaseModel.computeRs( pressure );
+    
+    // GAS
+    const double & gasSurfaceMoleDensity = m_gasPhaseModel.getSurfaceMoleDensity();
+    const double & gasSurfaceMassDensity = m_gasPhaseModel.getSurfaceMassDensity();
+    const double rvSat = m_gasPhaseModel.computeRv( pressure );
+
+    // Phase State - Negative flash type
+    double const Ko = rvSat * ( oilSurfaceMoleDensity + gasSurfaceMoleDensity * rsSat ) / ( gasSurfaceMoleDensity + oilSurfaceMoleDensity * rvSat );
+    double const Kg = ( oilSurfaceMoleDensity + gasSurfaceMoleDensity * rsSat ) / ( rsSat * ( gasSurfaceMoleDensity + oilSurfaceMoleDensity * rvSat ) );
+    double const V = zo / ( 1. - Kg ) + zg / ( 1. - Ko );
+  
+    if( ( 0 < V ) && ( V < 1 ) )  //Two-phase or both oil and gas saturated
```

Comment:
So you do not compute `LnFugacities` in this case?
Is it intended?

Reply:
So you can define those values in the constructor of `BlackOilFlashMultiphaseSystemProperties` like it's done for `DeadOilFlashMultiphaseSystemProperties`.

Reply:
Done